### ISSUE TITLE
CSRs for mms-agents are constantly being deleted recreated when x.509 authentication is enabled

### ISSUE THREAD

**Issue by andrii-litvinov** [mongodb/mongodb-enterprise-kubernetes#26]
I am working on enabling x.509 certificate authentication in the cluster and experience very weird issue that existing certificate signing requests being deleted and re-requested.

My cluster is created by Rancher 2.x in Hyper-V VMs on my workstation. I deploy sharded cluster with 1 shard, 3 `mongod`s, 3 configs and 2 `mongos`'.

Once mdb object is created I approve CSRs and cluster gets deployed. I see following output in the console:

```bash
kubectl get csr
NAME                           AGE     REQUESTOR                                                   CONDITION
mms-automation-agent.mongodb   4h47m   system:serviceaccount:mongodb:mongodb-enterprise-operator   Approved,Issued
mms-backup-agent.mongodb       4h47m   system:serviceaccount:mongodb:mongodb-enterprise-operator   Approved,Issued
mms-monitoring-agent.mongodb   4h47m   system:serviceaccount:mongodb:mongodb-enterprise-operator   Approved,Issued
```

Then I create see CSRs for internal TLS authentication between nodes in the cluster, I approve them too. I create and approve client certificate that I use connect to the cluster all works fine. But then suddenly see that `mms-*-agent.mongodb` are in Pending state again and often all other CSRs are no longer exist as well. Automation agent is not able to perform any changes to the cluster and and monitoring agent cannot report cluster state to Cloud Manager. So I have to manually approve new certificates again.

```bash
kubectl get csr
NAME                           AGE     REQUESTOR                                                   CONDITION
mms-automation-agent.mongodb   4h47m   system:serviceaccount:mongodb:mongodb-enterprise-operator   Pending
mms-backup-agent.mongodb       4h47m   system:serviceaccount:mongodb:mongodb-enterprise-operator   Pending
mms-monitoring-agent.mongodb   4h47m   system:serviceaccount:mongodb:mongodb-enterprise-operator   Pending
```

In operator logs I only see the CSRs were re-created, but nothing about the reason why existing CSRs were removed. I cannot track down the reason for this issue. Why does operator remove certificates and why does it remove certificated that it didn't create? Once I could correlate this issue with CPU thermal throttling, certs were removed about same time. Another case was when my public IP changed and I had to reconfigure white list in Ops Manager. But I cannot correlate other cases to any events in my environment. What can cause this issue and how can it be avoided?

**Comment by antonlisovenko** [comment]
Hi @andrii-litvinov 

Thanks for this report, we are investigating the issue from our side.

Meanwhile a couple of questions:
* Could you please clarify when do the agent certificates go to `Pending` state - right after creation or eventually? 
* Do they go to `Pending` always on the same time after creation? 
* Could you find any Kubernetes events about certificates - may be the system affects the work somehow?

**Comment by andrii-litvinov** [comment]
Hi Anton,

Thank you for a quick reply!

>Which version of the Operator and Kubernetes cluster you are using?

The Operator version is 1.1, k8s cluster version is v1.14.3.

>Could you please clarify when do the agent certificates go to Pending state - right after creation or eventually?

Certs go to Pending state some time after creation.

>Do they go to Pending always on the same time after creation?

The time is different. For instance this morning I started VMs and from the saved state and certs were re-requested after about an hour of running the cluster.

>Could you find any Kubernetes events about certificates - may be the system affects the work somehow?

I am still new to k8s, I tried to search where I can look for before posting this issue, but couldn't find where too look. I tried following:

```bash
$ kubectl get events --all-namespaces
No resources found.
```

**Comment by andrii-litvinov** [comment]
Earlier today I first approved monitoring cert and then after some time two others. And now I see following picture:

```bash
$ kubectl get csr
NAME                           AGE     REQUESTOR                                                   CONDITION
mms-automation-agent.mongodb   123m    system:serviceaccount:mongodb:mongodb-enterprise-operator   Approved,Issued
mms-backup-agent.mongodb       7m12s   system:serviceaccount:mongodb:mongodb-enterprise-operator   Pending
mms-monitoring-agent.mongodb   123m    system:serviceaccount:mongodb:mongodb-enterprise-operator   Approved,Issued
```
So it does seem that something in the cluster removed CSRs. Is it something in MongoDB Operator or k8s-specific? How can it be verified?

**Comment by andrii-litvinov** [comment]
I think I have found the reason. There is a certificate cleaner controller https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/certificates/cleaner/cleaner.go that runs every hour and deletes approved certs that were approved more than an hour ago. The time for deletion is different every time, that's why it was hard for me to correlate it to some scheduled task.

So I guess certificate auto-approval has to be configured in the cluster. I will look into it now. Or maybe it is better for the Operator to store approved certificates somewhere instead of requesting them each time from the API? In this case no auto-approval will be required. At least I think it is worth mentioning in the docs.

**Comment by antonlisovenko** [comment]
thanks @andrii-litvinov for the report and for this information about the cleaner
We found the bug in our code, and working on solution, I believe the hot fix will be released soon - we'll update this issue then

**Comment by antonlisovenko** [comment]
Yes, the issue was fixed in 1.2.1, closing the issue - feel free to reopen if it still occurs
